compiler/src/iree/compiler/DispatchCreation/SplitReduction.cpp

```diff
@@ -24,14 +25,344 @@ namespace mlir::iree_compiler::DispatchCreation {
+}
+
+// Matches the combiner pattern in a linalg.generic argmax-style reduction:
+// Example MLIR:
+// %4:2 = linalg.generic {
+//     indexing_maps = [...],
+//     iterator_types = ["parallel", "reduction"]
+//   } ins(%arg0 : tensor<?x128xbf16>) outs(%1, %3 : tensor<?xbf16>,
+//   tensor<?xi64>) {
+// ^bb0(%in: bf16, %out: bf16, %out_0: i64):
+//   %5 = linalg.index 1 : index
+//   %6 = arith.index_cast %5 : index to i64
+//   %7 = arith.maximumf %in, %out : bf16
+//   %8 = arith.cmpf ogt, %in, %out : bf16
+//   %9 = arith.select %8, %6, %out_0 : i64
+//   linalg.yield %7, %9 : bf16, i64
+// } -> (tensor<?xbf16>, tensor<?xi64>)
+//
+// This function extracts the `arith.maximumf`, `arith.cmpf`, and `arith.select`
+// operations from the body to facilitate transformations such as split
+// reduction.
+static LogicalResult
+collectArgmaxCombinerOps(linalg::GenericOp genericOp,
+                         SmallVector<Operation *, 4> &combinerOps) {
+  assert(IREE::LinalgExt::isArgmaxOp(genericOp) &&
+         "expected operation to be an argmax op");
+
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+
+  // Extract max value producer: arith.maximumf.
+  Value maxResult = yieldOp.getOperand(0);
+  auto maxOp = dyn_cast<arith::MaximumFOp>(maxResult.getDefiningOp());
+  if (!maxOp)
+    return failure();
+
+  // Extract index result producer: arith.select.
+  Value indexResult = yieldOp.getOperand(1);
+  auto selectOp = dyn_cast<arith::SelectOp>(indexResult.getDefiningOp());
+  if (!selectOp)
+    return failure();
+
+  // Extract the condition of the select, expected to be arith.cmpf with
+  // predicate OGT.
+  auto cmpOp = dyn_cast<arith::CmpFOp>(selectOp.getCondition().getDefiningOp());
+  if (!cmpOp || cmpOp.getPredicate() != arith::CmpFPredicate::OGT)
+    return failure();
+
+  combinerOps.push_back(maxOp);
+  combinerOps.push_back(selectOp);
+  combinerOps.push_back(cmpOp);
+
+  return success();
+}
+
+template <>
+FailureOr<linalg::SplitReductionResult> splitReductionImpl<linalg::GenericOp>(
+    RewriterBase &rewriter, linalg::GenericOp genericOp,
+    linalg::ControlSplitReductionFn controlSplitReductionFn) {
+  assert(IREE::LinalgExt::isArgmaxOp(genericOp) &&
+         "expected operation to be an argmax op");
+
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(genericOp);
+  Location loc = genericOp->getLoc();
+
+  linalg::SplitReductionOptions control = controlSplitReductionFn(genericOp);
+  int64_t ratio = control.ratio;
+  unsigned insertSplitIndex = control.index;
+  unsigned insertSplitDimension = control.index;
+  if (ratio <= 1) {
+    return rewriter.notifyMatchFailure(
+        genericOp, "split ratio needs to be greater than 1");
+  }
+
+  SmallVector<unsigned> dims;
+  genericOp.getReductionDims(dims);
+
+  if (dims.size() != 1) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "needs a single reduction dimension");
+  }
```

Comment:
I don't think you need this check again. This has already been checked by the isArgMaxOp helper.